lib/spack/docs/tutorial_modules.rst

```diff
@@ -922,63 +1274,255 @@ can add an arbitrary list of virtual providers to the triplet
+  root@module-file-tutorial:/# module load netlib-lapack
 
-  Lmod is automatically replacing "openblas/0.2.19" with "netlib-lapack/3.6.1"
+  Lmod is automatically replacing "openblas/0.2.20" with "netlib-lapack/3.6.1".
 
 
   Inactive Modules:
     1) py-numpy
 
-  Due to MODULEPATH changes the following have been reloaded:
+  Due to MODULEPATH changes, the following have been reloaded:
     1) netlib-scalapack/2.0.2
 
-making the use of tags to differentiate them unnecessary.
-Note that because we only compiled ``py-numpy`` with ``openblas`` the module
+Because we only compiled ``py-numpy`` with ``openblas`` the module
 is made inactive when we switch the ``LAPACK`` provider. The user
-environment will now be consistent by design!
+environment is now consistent by design!
+
+----------------------
+Working with templates
+----------------------
+
+As briefly mentioned in the introduction, Spack uses `Jinja2 <http://jinja.pocoo.org/docs/2.9/>`_
+to generate each individual module file. This means that you have at your disposal
+all the flexibility that a template engine may give when it comes to customizing what
+gets generated!
+
+^^^^^^^^^^^^^^^^^^^^^
+Module file templates
+^^^^^^^^^^^^^^^^^^^^^
+
+The templates that Spack uses to generate module files are stored in the
+``templates/module`` directory, and they all share the same common structure.
+Usually, they start with a header that identifies the type of
+module being generated. In the case of hierarchical module files it's:
+
+.. literalinclude:: ../../../templates/modules/modulefile.lua
+  :language: jinja
+  :lines: 1-6
+
+The statements within double curly brackets ``{{ ... }}`` denote
+`expressions <http://jinja.pocoo.org/docs/2.9/templates/#expressions>`_
+that will be evaluated and substituted at module generation time.
+The rest of the file is then divided into
+`blocks <http://jinja.pocoo.org/docs/2.9/templates/#template-inheritance>`_
+that can be overridden or extended by users, if need be.
+`Control structures <http://jinja.pocoo.org/docs/2.9/templates/#list-of-control-structures>`_
+, delimited by ``{% ... %}``,
+are also permitted in the template language:
+
+.. literalinclude:: ../../../templates/modules/modulefile.lua
+  :language: jinja
+  :lines: 73-88
+
+The locations where Spack looks for templates are specified
+in ``config.yaml``:
+
+.. literalinclude:: ../../../etc/spack/defaults/config.yaml
+  :language: yaml
+  :lines: 21-24
+
+and can be extended by users to employ custom templates, as we'll see next.
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Extend the default templates
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Let's assume one of our software is protected by group membership:
+allowed users belong to the same linux group, and access is granted at group level.
+Wouldn't it be nice if people that are not
+yet entitled to use it could receive an helpful message at module load time
+that tells them who to contact in your organization to be inserted in the group?
```

Comment:
interesting possibilities!
one could also do the same for modules of specific architectures in which an architecture represents a queue of a cluster... :)

Reply:
Sure. Maybe when we have more experience on the use cases we could collect them in some other section of the docs (not the tutorial). The idea here is that with templates you can easily add arbitrary code to module files.